Python wrapper for starSpace

freakeinstein opened the pull request:
Hi, I have wrote a python wrapper for starSpace. It uses pybind11 can be installed with conan package manager (conan is required in the dev. env.). To build wrapper library, CMake is required. 

`build.sh` inside python directory will build the wrapper and run test script.

freakeinstein commented:
@ledw Sure. I found one issue as well. The program can't be terminated with SIGINT when training is started. I think, C++ thread is not considering SIGINT event. Need to look into it. Any help appreciated.

jamesjrg commented:
Thanks for the work, FWIW I had the following minor issues:

1. Making the starspace binary dependent on libstarspace.so means you have to set LD_LIBRARY_PATH or some such, maybe it should be statically linked? 

2. Maybe the default makefile target should remain the main starspace binary?

3. It would be nice if the bindings included e.g. getNgramVector, getDocVector and maybe predictOne, though I guess this would also require a numpy binding for the matrix stuff so isn't just a matter of adding a line for each extra function

freakeinstein commented:
> Maybe the default makefile target should remain the main starspace binary?

@jamesjrg That's a good suggestion. We can keep a separate makefile for those who needed python binding. Right?

>It would be nice if the bindings included e.g. getNgramVector, getDocVector and maybe predictOne, though I guess this would also require a numpy binding for the matrix stuff so isn't just a matter of adding a line for each extra function

I'll try to do it. And you are welcome if you could help.

jamesjrg commented:
Looks like numpy binding is done with the def_buffer function:

https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html
https://github.com/facebookresearch/fastText/blob/master/python/fastText/pybind/fasttext_pybind.cc

If noone else has done it in like a month's time I might well take a look, but a bit busy right now and quite likely someone else will get there first!

freakeinstein commented:
UPDATE: Incorporated @jamesjrg 's first two suggestions
-------------------------------------------------------------------------
1. Reduced dependency from two to one. Only starwrap.so is needed.
2. added separate makefile for python bindings (makefile_py), reverted original makefile.

freakeinstein commented:
**Update:** I have incorporated `getDocVector` numpy bindings. I didn't bind `getNgramVector`, because of the return type (I can't find a way to cast `MatrixRow` typedef into a python object). I would appreciate if someone could help. The remaining part is a python example as @ledw suggested.

rnditdev commented:
[This](https://github.com/inducer/boost-python-ublas-subset/tree/master/boost_subset) has a wrapper of ublas matrix_row, which is the original in the typedef of Starspace, hope it helps, don't have time to invest in solving it right now. Thanks for your work.

ledw commented on python/build.sh:
```diff
@@ -0,0 +1,29 @@
+cmake --build .
+cd -
+
+echo "#############################  run test ############################# "
```
Could you add comments here about what the expected results are from running the test?

freakeinstein replied:
added.

rnditdev commented:
I'd also needing to have additional methods to the bindings, StarSpace::predictOne, how difficult do you estimate will the work be for this and a couple related methods? Thanks!

freakeinstein commented:
@ledw Update: Added comments in build script, Added README for python, added reference from root README. 

@rnditdev I'm sorry for the late reply, suddenly got into unbalanced mode. `predictOne` has also got typedefs as params, that I have no idea on how to bind.

rnditdev commented:
Could we simplify the interface of PredictOne, so we facilitate the creation of bindings?

nickyongzhang commented:
Thank you for the python wrapper again. But I am wondering whether it's possible to return the responses as objects rather than just print it on the screen. For example, when calling `sp.nearestNeighbor('some text',10)` in the `test.py` file, I think it's better we can assign the response to a variable `neighbors=sp.nearestNeighbor('some text',10)` so that we can do better downstream analysis. Similarly, it's better we can record the training history into a history object. 
Sorry that I am not quite familiar with c++ projects and may not able to help, but I do wish someone can help make the python wrapper more user-friendly.
